### PR TITLE
[MPS] fix rand and uniform_ upper-bound rounding

### LINKED ISSUE #197079: [MPS] torch.rand and uniform_ return the exclusive upper bound for float16 and bfloat16
### 🐛 Describe the bug

On MPS, `torch.rand` and `Tensor.uniform_` in `float16` and `bfloat16` return the exclusive upper bound. [`torch.rand`](https://pytorch.org/docs/stable/generated/torch.rand.html) documents the half-open interval `[0, 1)`. float32 on MPS, and every dtype on CPU, never return it.

```python
import torch

torch.manual_seed(0)
n = 10_000_000
for device in ("cpu", "mps"):
    for dtype in (torch.float32, torch.float16, torch.bfloat16):
        x = torch.rand(n, dtype=dtype, device=device)
        y = torch.empty(n, dtype=dtype, device=device).uniform_(0, 4)
        print(f"{device:4s} {str(dtype):15s} rand==1: {int((x == 1).sum()):6d}   uniform_(0, 4)==4: {int((y == 4).sum()):6d}")
```

Output. torch 2.14.0 and nightly 2.15.0.dev20260913 print exactly the same counts:

```
cpu  torch.float32   rand==1:      0   uniform_(0, 4)==4:      0
cpu  torch.float16   rand==1:      0   uniform_(0, 4)==4:      0
cpu  torch.bfloat16  rand==1:      0   uniform_(0, 4)==4:      0
mps  torch.float32   rand==1:      0   uniform_(0, 4)==4:      0
mps  torch.float16   rand==1:   2425   uniform_(0, 4)==4:   2432
mps  torch.bfloat16  rand==1:  19227   uniform_(0, 4)==4:  19449
```

The rates are about 2^-12 for float16 and about 2^-9 for bfloat16. Those are the widths of `[1 - ULP/2, 1)` in each dtype, which is the band of float32 values that rounds to `1.0` when cast. That fits the current kernel in `aten/src/ATen/native/mps/kernels/Distributions.metal`, which computes the value in `float` and casts afterwards:

```metal
float u = c10::metal::detail::uint32_to_uniform_float(raw[i]);
output[base + i] = static_cast<T>(from + scale * u);
```

This is the MPS counterpart of #96947, which was the same bug on CUDA, fixed by #96962. #190931 ("[MPS] Fix uniform upper-bound rounding") proposed the same fix for MPS but was closed on 2026-08-04 without merging. I found no issue that tracks the bug, and `Distributions.metal` on `main` still has the lines above.

**Impact.** Code that truncates a uniform draw to an index gets an out-of-range index. In kornia, `Uniform(0, N)` draws followed by `.long()` or `floor` give `index N is out of bounds` errors in a color augmentation, and crop and erase boxes that extend one pixel past the image edge, on MPS in half precision only. `torch.log(1 - torch.rand(...))` also returns `-inf`, as the #190931 description shows.

### Versions

- torch 2.14.0 and 2.15.0.dev20260913 (nightly)
- Python 3.11
- macOS 26.5.1, Apple M1 (8 GB), MPS backend; `PYTORCH_ENABLE_MPS_FALLBACK` unset

Posted on behalf of @ducha-aiki by Claude (Opus 5).

cc @pbelevich @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by Isalia20**
By adding `value == to ? from : value` ternary at the end
Use `volatile` to prevent erroneous elimination of `static_cast<bfloat>` see reproducer below
```python
import torch

lib = torch.mps.compile_shader("""
#include <metal_stdlib>
using namespace metal;


bfloat volcast(volatile float x) {
  return static_cast<bfloat>(x);
}

kernel void repro_bug(device bfloat* out,
                  constant float2& bounds,
                  constant float* samples,
                  uint i [[thread_position_in_grid]]) {
  bfloat value = bfloat(bounds.x + (bounds.y - bounds.x) * samples[i]);
  out[i] = value == bounds.y ? static_cast<bfloat>(bounds.x) : value;
}

kernel void repro_works(device bfloat* out,
                  constant float2& bounds,
                  constant float* samples,
                  uint i [[thread_position_in_grid]]) {
  bfloat value = bfloat(bounds.x + (bounds.y - bounds.x) * samples[i]);
  out[i] = value == bounds.y ? volcast(bounds.x) : value;
}
""")

samples = torch.tensor([0.9999], device="mps")
out_bug = torch.empty(1, device="mps", dtype=torch.bfloat16)
out_works = torch.empty(1, device="mps", dtype=torch.bfloat16)
for low, high in [(0.0, 1.0), (1.0, 2.0), (-4.0, -2.0)]:
    bounds = torch.tensor([low, high], device="mps")
    lib.repro_bug(out_bug, bounds, samples, threads=1)
    lib.repro_works(out_works, bounds, samples, threads=1)
    print(f"bounds={[low, high]}: expected {low}, got from buggy cast {out_bug.item()} from volatile cast {out_works.item()}")
```

Submitted ticket with Feedback Assistant, ticket number FB24789985

Fixes #197079

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/Distributions.metal [hunk 1]
```diff
@@ -122,7 +122,10 @@ kernel void uniform_dist(
   uint count = min(4u, numel - base);
   for (uint i = 0; i < count; ++i) {
     float u = c10::metal::detail::uint32_to_uniform_float(raw[i]);
-    output[base + i] = static_cast<T>(from + scale * u);
+    T value = static_cast<T>(from + scale * u);
+    // Casting to T can round up to the excluded upper bound.
+    output[base + i] =
+        static_cast<T>(float(value) == params.w ? params.z : float(value));
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/Distributions.metal]: This one will not work for bfloat16 I assume
```suggestion
        float(value) == params.w ? static_cast<T>(params.z) : value;
```

**Review by malfet — DISMISSED** [review]

**Review by malfet — APPROVED** [review]
LGTM, though not sure why you need to pass from and two twice

File: aten/src/ATen/native/mps/operations/Distributions.mm [hunk 2]
```diff
@@ -226,7 +226,14 @@ static void bernoulli_tensor_kernel_mps(const TensorBase& self, const TensorBase
 REGISTER_MPS_DISPATCH(bernoulli_tensor_stub, &bernoulli_tensor_kernel_mps)
 
 static void uniform_kernel_mps(TensorIteratorBase& iter, double from, double to, std::optional<Generator> gen) {
-  distribution_kernel_mps_impl(iter, from, to, "uniform_dist", 1, gen);
+  AT_DISPATCH_FLOATING_TYPES_AND2(kHalf, kBFloat16, iter.dtype(), "uniform_kernel_mps", [&] {
+    // Narrow bounds on the host to avoid Metal miscompiling constant float-to-bfloat casts.
+    std::array<float, 4> params{static_cast<float>(from),
+                                static_cast<float>(to),
+                                static_cast<float>(static_cast<scalar_t>(from)),
+                                static_cast<float>(static_cast<scalar_t>(to))};
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Distributions.mm]: Can you explain why do you need to do it on the host rather than on device?

Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/Distributions.mm]: See comment above this line, on line 230

Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/Distributions.mm]: repro
```
import torch

lib = torch.mps.compile_shader("""
#include <metal_stdlib>
using namespace metal;

kernel void repro(device bfloat* out, 
                  constant float2& bounds, 
                  constant float* samples,
                  uint i [[thread_position_in_grid]]) {
  bfloat value = bfloat(bounds.x + (bounds.y - bounds.x) * samples[i]);
  out[i] = value == bfloat(bounds.y) ? bfloat(bounds.x) : value;
}
""")

bounds = torch.tensor([0.0, 1.0], device="mps")
samples = torch.tensor([0.9999], device="mps")
out = torch.empty(1, device="mps", dtype=torch.bfloat16)
lib.repro(out, bounds, samples, threads=1)
print("Expected [0.0], got", out.cpu().tolist())
```

File: aten/src/ATen/native/mps/kernels/Distributions.metal (hunk 1 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/kernels/Distributions.metal]: Original passes as well because bf16 -> fp32 is representeed without loss, but round trip yeah is redundant, will fix
